### DISCUSSION TITLE
Getting I/O errors - probably problem between chair and the keyboard

### DISCUSSION THREAD

**Discussion by darmach** [jpcornil-git/HA-mcp23017#5] [General]
Hi, I have HA OS installed on Rpi3B, and MCP23017 connected to i2c-1 bus (if I didn't mix something up - connected data pins to those right next to 5V and GND on Rpi3B GPIO)

With HA-mcp23017 installed via HACS I do get errors below in the logs:

```
2022-02-28 23:46:12 ERROR (SyncWorker_5) [custom_components.mcp23017] Unable to access mcp23017-0x23 ([Errno 121] Remote I/O error)
2022-02-28 23:46:13 ERROR (SyncWorker_5) [custom_components.mcp23017] Unable to access mcp23017-0x23 ([Errno 121] Remote I/O error)
2022-02-28 23:46:13 ERROR (SyncWorker_3) [custom_components.mcp23017] Unable to access mcp23017-0x23 ([Errno 121] Remote I/O error)
2022-02-28 23:46:15 ERROR (SyncWorker_5) [custom_components.mcp23017] Unable to access mcp23017-0x23 ([Errno 121] Remote I/O error)
```

Tried to test this with apk installed i2c-tools:
```
[core-ssh ~]$ apk add i2c-tools
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/main/aarch64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/community/aarch64/APKINDEX.tar.gz
(1/1) Installing i2c-tools (4.2-r1)
Executing busybox-1.33.1-r6.trigger
OK: 90 MiB in 100 packages
[core-ssh ~]$ ls -l /dev/i2c-1
crw-------    1 root     root       89,   1 Oct 12 17:24 /dev/i2c-1
[core-ssh ~]$ i2cdetect -y 1
Error: Could not open file `/dev/i2c-1': Operation not permitted
[core-ssh ~]$ whoami
root
```

I suspect I don't have access to this device because of container separation (this was executed in HA ssh, so in ssh addon container I guess?)

Can you advise what I might be doing wrong?

**Comment by jpcornil-git** [comment]
Hi @darmach,

I'm not used to HAOS (using a core install) but your issue is definitively related to access rights to the i2c bus within the container running HA, cfr. [L188](https://github.com/jpcornil-git/HA-mcp23017/blob/main/custom_components/mcp23017/__init__.py#L188) (and ssh). A few questions:

- How did you enabled i2c-1 on your platform ? Using [HassOS I2C Configurator I2C](https://community.home-assistant.io/t/add-on-hassos-i2c-configurator/264167) ?
- Does i2cdetect give the same result for i2c-0 ?
- Are i2c drivers loaded et i2c enabled ? (i2c_bcmxx may differ, not sure about location of config.txt on HAOS but based on [HassOS I2C Configurator script](https://github.com/adamoutler/HassOSConfigurator/blob/main/Pi4EnableI2C/run.sh#L54) it is in /tmp/$partition/config.txt. Dump of mine (RPi OS):

```bash
pi@RPiv4:~ $ grep i2c /boot/config.txt
dtparam=i2c_arm=on
pi@RPiv4:~ $ lsmod | grep i2c
i2c_bcm2835            16384  2
i2c_dev                20480  4
```

I can give a try to HAOS using a spare RPi4B or a QEMU machine but not before next week.

**Comment by darmach** [comment]
Thanks for the response @jpcornil-git !

I enabled it manually, logging on physical console as root to get the access to /mnt/boot

In config.txt I enabled only i2c_arm - I wanted to get only the i2c-1 header active, the one (I hope I found that right) next to 5V and GND pins.

I can see i2c modules loaded:
```
[core-ssh ~]$ lsmod | grep i2c
i2c_dev                20480  8
i2c_bcm2835            16384  4
```
(Hmm, but that's a suspiciously high device count for Rpi3, right? 8 and 4?)

Also the device is visible in /dev:
```
[core-ssh ~]$ ls -l /dev/ | grep i2c
crw-rw-rw-    1 root     root       89,   1 Oct 12 17:24 i2c-1
```

No hurry at all, if you could find some time next week to test this it would be awesome!

**Comment by darmach** [comment]
Hi @jpcornil-git, I tested this a bit today, ran a `--privileged` ubuntu container on the side, and was able to `i2cdetect` the addres without problem.

Looks like the root cause is HA container privilege separation, can we somehow safely (or not) work around it while still using HASSOS? Else I will have to think about migration to Some Rpi linux distribution and running HA on top of that...

**Comment by jpcornil-git** [comment]
Note that I see [here](https://github.com/home-assistant/operating-system/tree/dev/Documentation/boards/raspberrypi) that dtparam=i2c1=on and dtparam=i2c_arm=on should be added to config.txt

**Comment by jpcornil-git** [comment]
From what I can see HA OS supervisor is running homeassistant container in privileged mode already, cfr. https://github.com/home-assistant/supervisor/blob/main/supervisor/docker/homeassistant.py#L133
=> likely only a matter of enabling i2c in HA OS as described above

**Comment by darmach** [comment]
@jpcornil-git Thanks for moving this into a discussion! I can confirm that you're right, ha is already privileged, found the culprit - in the end the root cause were cheap cables... Which ended my MCP23017 to get 2.1V as VCC.

**Comment by jpcornil-git** [comment]
Great to hear this (issue was related to accessing the device then not the bus as I thought based on i2cdetect result)
-> everything/mcp23017 integration working for you now ?

**Comment by darmach** [comment]
Yes, integration works!

What's interesting (leaving this for anyone fightning with i2c on Rpi under HASSOS) while I can properly i2cdetect address of the device from separate container (plain ubuntu) running on HASSOS, still if I try to execute `i2cdetect -y 1` inside homeassistant container, I get access denied:

```
[core-ssh ~]$ i2cdetect -y 1
Error: Could not open file `/dev/i2c-1': Operation not permitted
```

That's weird, especially when `homeassistant` container is indeed a priviledged one:

```
# docker inspect homeassistant | grep -i privil
            "Privileged": true,
```

Also - if you need only one i2c bus (as you can address more devices on the same bus) then on Rpi3B it's enough to enable:
`dtparam=i2c_arm=on`

This enables the i2c-1 bus, you can verify it worked when you see `/dev/i2c-1` created after the reboot. That's the i2c header right next to 5V, GND and 3.3V pins on the GPIO

**Comment by jpcornil-git** [comment]
Just installed HAOS on a spare RPi (without i2c devices attached)
-> You should run i2cdetect from homeassistant container (iso addon_core_ssh)
```
# docker ps
CONTAINER ID   IMAGE                                                           COMMAND                  CREATED          STATUS          PORTS                                   NAMES
8820bb0770ab   homeassistant/aarch64-addon-ssh:9.3.0                           "/init"                  29 minutes ago   Up 29 minutes   0.0.0.0:22->22/tcp, :::22->22/tcp       addon_core_ssh
5311519de6f0   ghcr.io/home-assistant/aarch64-hassio-multicast:2021.04.0       "/init"                  29 minutes ago   Up 29 minutes                                           hassio_multicast
2e1dee19655c   ghcr.io/home-assistant/aarch64-hassio-audio:2021.07.0           "/init"                  29 minutes ago   Up 29 minutes                                           hassio_audio
dd48c5cfbf65   ghcr.io/home-assistant/aarch64-hassio-dns:2021.06.0             "/init"                  29 minutes ago   Up 29 minutes                                           hassio_dns
7a4fa51a33cd   ghcr.io/home-assistant/aarch64-hassio-cli:2021.12.0             "/init /bin/bash -c …"   29 minutes ago   Up 29 minutes                                           hassio_cli
27b1c220a9e0   ghcr.io/home-assistant/raspberrypi4-64-homeassistant:2022.2.9   "/init"                  9 hours ago      Up 29 minutes                                           homeassistant
c87217239505   ghcr.io/home-assistant/aarch64-hassio-observer:2021.10.0        "/init"                  9 hours ago      Up 29 minutes   0.0.0.0:4357->80/tcp, :::4357->80/tcp   hassio_observer
0f45ca8a5f34   homeassistant/aarch64-hassio-supervisor:latest                  "/init"                  9 hours ago      Up 29 minutes                                           hassio_supervisor
```

```bash
# docker exec -it homeassistant /bin/bash
bash-5.1# apk add i2c-tools
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/main/aarch64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.14/community/aarch64/APKINDEX.tar.gz
(1/1) Installing i2c-tools (4.2-r1)
Executing busybox-1.33.1-r3.trigger
OK: 156 MiB in 181 packages
bash-5.1# i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- -- 
```
